yt_dlp/cookies.py

```diff
@@ -577,42 +602,183 @@ def parse_safari_cookies(data, jar=None, logger=YDLLogger()):
+                return _LinuxDesktopEnvironment.GNOME
+            else:
+                return _LinuxDesktopEnvironment.UNITY
+        elif xdg_current_desktop == 'GNOME':
+            return _LinuxDesktopEnvironment.GNOME
+        elif xdg_current_desktop == 'X-Cinnamon':
+            return _LinuxDesktopEnvironment.CINNAMON
+        elif xdg_current_desktop == 'KDE':
+            return _LinuxDesktopEnvironment.KDE
+        elif xdg_current_desktop == 'Pantheon':
+            return _LinuxDesktopEnvironment.PANTHEON
+        elif xdg_current_desktop == 'XFCE':
+            return _LinuxDesktopEnvironment.XFCE
+    elif desktop_session is not None:
+        if desktop_session in ('mate', 'gnome'):
+            return _LinuxDesktopEnvironment.GNOME
+        elif 'kde' in desktop_session:
+            return _LinuxDesktopEnvironment.KDE
+        elif 'xfce' in desktop_session:
+            return _LinuxDesktopEnvironment.XFCE
     else:
-        logger.debug('using find-generic-password to obtain password')
-        proc = Popen(
-            ['security', 'find-generic-password',
-             '-w',  # write password to stdout
-             '-a', browser_keyring_name,  # match 'account'
-             '-s', '{} Safe Storage'.format(browser_keyring_name)],  # match 'service'
-            stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
-        try:
-            stdout, stderr = proc.communicate_or_kill()
-            if stdout[-1:] == b'\n':
-                stdout = stdout[:-1]
-            return stdout
-        except BaseException as e:
-            logger.warning(f'exception running find-generic-password: {type(e).__name__}({e})')
-            return None
+        if 'GNOME_DESKTOP_SESSION_ID' in env:
+            return _LinuxDesktopEnvironment.GNOME
+        elif 'KDE_FULL_SESSION' in env:
+            return _LinuxDesktopEnvironment.KDE
+        else:
+            return _LinuxDesktopEnvironment.OTHER
+
+
+def _choose_linux_keyring(logger):
+    """
+    https://chromium.googlesource.com/chromium/src/+/refs/heads/main/components/os_crypt/key_storage_util_linux.cc
+    SelectBackend
+    """
+    desktop_environment = _get_linux_desktop_environment(os.environ)
+    logger.debug('detected desktop environment: {}'.format(desktop_environment.name))
+    if desktop_environment == _LinuxDesktopEnvironment.KDE:
+        linux_keyring = _LinuxKeyring.KWallet
+    elif desktop_environment == _LinuxDesktopEnvironment.OTHER:
+        linux_keyring = _LinuxKeyring.BasicText
+    else:
+        linux_keyring = _LinuxKeyring.GnomeKeyring
+    return linux_keyring
+
+
+def _get_kwallet_password(browser_keyring_name, logger):
+    if shutil.which('kwallet-query') is None:
+        logger.error('kwallet-query command not found. Cannot read from kwallet')
+        return b''
+
+    logger.debug('using kwallet-query to obtain password from kwallet')
+    proc = Popen(['kwallet-query',
```

Comment:
is this guarenteed to be present in relevent systems?

Reply:
`/usr/bin/kwallet-query` is provided by the `kwallet` package on arch (which keeps it's packages minimal) (https://archlinux.org/packages/extra/x86_64/kwallet/files/)
so I would assume all but the most exotic systems with kwallet installed would have `kwallet-query`. But regardless I should move `Popen` into the try/catch.

Hopefully the error that would be printed is clear enough to figure out how to resolve the problem:
```
[debug] using kwallet-query to obtain password from kwallet
WARNING: exception running kwallet-query: FileNotFoundError([Errno 2] No such file or directory: 'kwallet-query')
```
but if you prefer, I could put `if shutil.which('kwallet-query') is None: <print a specific error message>`

Reply:
the default error is fine imo